### ISSUE TITLE
Race condition results in Android replay server behaving like a targeted app

### ISSUE THREAD

**Issue by jcortell68** [baldurk/renderdoc#2651]
## Description
The RenderDoc Android replay server is subject to a race condition on startup that results in its libVkLayer_GLES_RenderDoc.so thinking it’s *not* running in the replay server. The library thus acts like it’s in an app to be targeted/debugged and opens a control socket (38920) instead of a replay server socket (39920). The most obvious symptom of this is that when you ask RenderDoc to launch an app for debugging, it displays a prompt that says something else is already being debugged (which you know is false). You can then go to File > Attach To Running Instance and you will see the replay server app (e.g., org.renderdoc.renderdoccmd.arm64) listed as something you can attach to, which should never happen.

This doesn’t happen on all devices, but it happens on our devices that have Android 12. A difference between the two situations is the native libraries that are loaded implicitly when the replay server APK starts. In the problematic case (our devices), libhwui.so is loaded immediately when the apk is started. In the working cases (other devices) libhwui.so isn’t loaded. That is, the same exact APK experiences different library loading in the working and failing environments. I’ll provide more details at the end.

![renderdoc_race_condition](https://user-images.githubusercontent.com/109228283/178758018-ce892049-653e-4b02-b793-548afb607aa3.png)

The reason libhwui being in the mix is problematic is that it launches a dedicated thread (1) to initialize the Vulkan/GLES environments for the process. http://aospxref.com/android-12.0.0_r3/xref/frameworks/base/libs/hwui/renderthread/RenderThread.cpp#187
This thread starts making vulkan calls (2), and libvulkan.so will, upon the first significant API call, initiate its layer loading. That logic looks for layer libraries in an apk’s lib directory unconditionally (i.e., no properties need to be set). See http://aospxref.com/android-12.0.0_r3/xref/frameworks/base/core/java/android/os/GraphicsEnvironment.java#331
That means that very shortly after the replay server apk starts running, libvulkan will load libVkLayer_GLES_RenderDoc.so (3).

Meanwhile, back on the main thread…the process is busy going other initialization after loading libhwui.so and execution eventually gets to the Loader java class in the replay server that does a `System.loadLibrary("renderdoccmd")` (4), which loads the replay server logic, all of which is C++ code built into librenderdoccmd.so.

If the main thread doesn’t get to that Java loadLibrary() call before the secondary libhwui thread exercises the vulkan layer loading, then we have a bad situation. The “DllMain” in libVkLayer_GLES_RenderDoc.so tries to determine if it’s in the replay server by doing a dlysm(“renderdoc__replay__marker”). That symbol exists in librenderdoccmd.so. If libVkLayer_GLES_RenderDoc.so was first loaded by the vulkan layer loading, then the library’s DllMain() ran at that time and failed, since librenderdoccmd.so hasn’t yet been loaded.

That dlsym(marker) approach works fine when libVkLayer_GLES_RenderDoc.so is loaded by the loading of renderdoccmd.so in Loader.java, which is how things are supposed to happen. But this secondary thread created by libhwui throws a wrench into things by creating the possibility that libVkLayer_GLES_RenderDoc.so gets loaded prematurely. For our devices it happens more often than not–anywhere from 75% to 95% of the time, depending on the device and build.

I did not find an elegant solution other than to replace the dlsym(marker) call with a check of the process executable and compare it to the name of the Android replay server. I conditionalized this as alternate logic for Android only, of course. If someone can think of a better solution, I’m all for it. I’ll submit the change as a pull request. In general though, it seems hazardous/fragile for libVkLayer_GLES_RenderDoc.so to assume the other library in the replay server (renderdoccmd.so) will be loaded first. Normally (desktop and embedded Linux), the replay server is just an executable and so the assumption is rock solid; the executable will of course be loaded prior to any library the executable dlopens.

Finally, I want to provide more details on why this affects our Android devices but not others I’ve tried (production Android phones). I mentioned early that I noticed a stark difference in the activity in the dynamic loader when launching the replay server APK. Here are some details, in case they provide clues.

When I enable tracing in the bionic loader of a production phone (lacks problem), I see that the first dlopen() in the replay server process is for `/system/framework/oat/arm64/org.apache.http.legacy.odex`, done by `/apex/com.android.art/lib64/libart.so`. On our devices (has problem), the first dlopen is for `libandroid.so` done by `/system/lib64/libhwui.so`. To enable loader tracing (to logcat), simply do 
   `adb shell setprop debug.ld.app.org.renderdoc.renderdoccmd.arm64 dlopen`

## Steps to reproduce
This happens on our devices but not production phones I've tried.
Also, this has nothing to do with any app. The problem happens entirely within the replay server apk.

For Android devices where the premature loading of libVkLayer_GLES_RenderDoc.so happens
1. Launch RenderDoc
2. Select the Android device in the replay server
3. Try to launch any app from RenderDoc. You'll get a dialog saying you are already debugging something (although you're not)
4. Open the Attach to Running Instance and you'll see the replay server listed as something you can attach to (which makes no sense).

## Environment
* RenderDoc version: 1.20
* Operating System: 
    Host: All
    Device: Android 12
* Graphics API: N/A since no app is involved.

**Comment by jcortell68** [comment]
Baldur, I'm posting my fix here for initial consideration, as it's small. If you think this is acceptable, I'll create a pull request.

```
diff --git a/renderdoc/os/posix/posix_libentry.cpp b/renderdoc/os/posix/posix_libentry.cpp
index 42756cb68..ee3bf453d 100644
--- a/renderdoc/os/posix/posix_libentry.cpp
+++ b/renderdoc/os/posix/posix_libentry.cpp
@@ -31,7 +31,16 @@ void ResetHookingEnvVars();
 // DllMain equivalent
 void library_loaded()
 {
-  if(LibraryHooks::Detect("renderdoc__replay__marker"))
+  bool isReplayServer = false;
+#if ENABLED(RDOC_ANDROID)
+  rdcstr exeFilename;
+  FileIO::GetExecutableFilename(exeFilename);
+  isReplayServer = exeFilename.beginsWith(RENDERDOC_ANDROID_PACKAGE_BASE);
+#else
+  isReplayServer = LibraryHooks::Detect("renderdoc__replay__marker");
+#endif
+
+  if(isReplayServer)
   {
     RDCDEBUG("Not creating hooks - in replay app");

```

**Comment by baldurk** [comment]
I genuinely hate Android so much, I've lost count of how many times I thought it couldn't surprise me with how completely and fundamentally terrible a platform it is and yet it still finds ways to shock me.

The root problem of this is that Android thinks it is entitled to load and run our code in arbitrary order, even before it knows whether vulkan is being used at all(?!). This also means presumably that the vulkan code needs to be thread-safe vs the server code which is not something I planned on guaranteeing since on every other platform it runs single threaded on replay. I honestly don't know what other things this might break.

The issue with your hack is that it assumes the replay will only happen within RenderDoc's named packages. This is most commonly the case but someone could create their own packages if e.g. if they're brave enough to try to use the replay API for automation on Android. I don't want to explicitly break that capability since that is intended to work. Indeed I originally used to just look at executable name until it became clear that that was _not_ sufficient to cover all cases, so I moved to symbol search which is more robust on all non-broken platforms.

Is there a way to block the loading of this libhwui.so, or defer it so that it runs at a more sensible time? Is there any way to load our libraries before that at an earlier point in initialisation? Can we prevent it from loading layers or otherwise configure its vulkan initialisation? The more reliable fix for this is to find a way to hack Android to run the code that we tell it to instead of the code it feels like executing, especially since that could change tomorrow and then we have a new problem.

Since we're looking at hacks, is there a reliable order that layers are loaded in? E.g. if it goes alphabetically we could rename `librenderdoccmd.so` to `libVkLayer_aaaa_renderdoccmd.so` and implement entry stubs that return nothing. Then at least we're in control of which code gets loaded first and we can load our own libraries in a sensible order.

**Comment by jcortell68** [comment]
> Is there a way to block the loading of this libhwui.so, or defer it so that it runs at a more sensible time? 

I can't imagine how. It's part of the blackbox logic that happens when the apk is launched. I don't think there are any knobs to that process. At the end of the day, there is no reference to that library in the replay server code.

> Is there any way to load our libraries before that at an earlier point in initialisation? 

Here, too...not that I know of. The existing loadLibrary logic is in a generated java file that is part of a cookie-cutter native activity solution for apks that are effectively native code.

> Can we prevent it from loading layers or otherwise configure its vulkan initialisation? 

Preventing the layer loading was what I was hoping for, but I studied libvulkan and I see no way to prevent it.
References:
http://aospxref.com/android-12.0.0_r3/xref/frameworks/native/vulkan/libvulkan/api.cpp#1186
http://aospxref.com/android-12.0.0_r3/xref/frameworks/native/vulkan/libvulkan/layers_extensions.cpp?fi=DiscoverLayers#DiscoverLayers

Also the opengrok reference I put in the issue description...it shows the apk's lib directory being added to the search path unconditionally. 

>  is there a reliable order that layers are loaded in? E.g. if it goes alphabetically ....

Unfortunately not. On Android (and presumably Linux), it uses opendir/readdir alone to iterate, and that API provides no guarantees on ordering.

> I originally used to just look at executable name until it became clear that that was not sufficient to cover all cases

Ah, ok. Well, that settles that. I can understand why you wouldn't want to go back to something that has already proven problematic. Maybe the proper solution here is to pursue this with Google/AOSP to see if they could provide a way in their vulkan loader to forego adding the apk's lib directory to the layer loading searchpath via a property instead of adding it unconditionally.

So, feel free to close this issue. I've captured/shared the problem here, which was honestly my primary objective. I suspected a proper fix was going to be tricky. I'll create a ticket with AOSP with the ask above and reference this issue. Maybe we get lucky and Android gives us a path for a proper solution down the road.

**Comment by baldurk** [comment]
My worry is that from what you're saying there's nothing specific about this - any Android device _could_ run into this, there's just something about your specific setup which makes it much more likely. With a broken platform like Android that's the kind of thing that can quickly snowball into "lots of people see this".

> Maybe the proper solution here is to pursue this with Google/AOSP to see if they could provide a way in their vulkan loader to forego adding the apk's lib directory to the layer loading searchpath via a property instead of adding it unconditionally.

I'm not sure what you mean by a property here? If we have code running early enough that can set some kind of flag, could we not instead use that code to load the library in the correct order?

I have zero confidence that any change will come from this from Google/AOSP, so I'm afraid we will need to do something about it. If there's no possible feasible hack then we should find a way to detect this case and disable the replay server indicating that this bug has occurred.

One other thing I considered, would it be feasible to extend the 'symbol detection' on android (in `LibraryHooks::Detect` in `android_hook.cpp`) to call into java and look for that symbol in the java side as well? E.g. require that a class of that name is exported somewhere. I don't know whether this code is running even before java is loaded though, and on top of that like I say there could be other things that break even if this is fixed because the replay server is _not_ intended to run on multiple threads simultaneously.

**Comment by baldurk** [comment]
Another idea that @Zorro666 suggested - can you place custom properties in an android manifest and retrieve them in `LibraryHooks::Detect` ?

**Comment by jcortell68** [comment]
> I'm not sure what you mean by a property here? If we have code running early enough that can set some kind of flag, could we not instead use that code to load the library in the correct order?

The layer search path used by the Android vulkan is controlled by Android properties, and then Android framework code reads those properties and extends the search path accordingly. The host side RenderDoc code sets those properties via adb. What's missing is a property that says "don't load any layers at all". If we had that, then we could set the property when we launch the replay server.

> One other thing I considered, would it be feasible to extend the 'symbol detection' on android (in LibraryHooks::Detect in android_hook.cpp) to call into java and look for that symbol in the java side as well?

Hm. Good idea. Let me look into that.

> Another idea that @Zorro666 suggested - can you place custom properties in an android manifest and retrieve them in LibraryHooks::Detect ?

I'm unaware of any, but I'll look into that, too

**Comment by baldurk** [comment]
I'm not an Android expert, but if you mean the ones set with `adb shell setprop` then IMO those should not be used as they are unreliable.

**Comment by jcortell68** [comment]
> I'm not an Android expert, but if you mean the ones set with `adb shell setprop` then IMO those should not be used as they are unreliable.

RenderDoc already has a deep reliance on setting properties that way.

**Comment by baldurk** [comment]
Yes unfortunately so, often mandated by the system only having those properties for necessary settings. That's how I know that the properties are unreliable because there are cases where it will break (e.g. people sometimes run into problems where properties aren't unset properly and no vulkan programs work anymore on the device).

**Comment by jcortell68** [comment]
I've looked into the latest two suggestions. They aren't feasible. Calling back into the Java world from the C++ world requires context that becomes available only after the Java world has first called into the C++ world. The formal C++ world of the apk is in renderdoccmd.so, which hasn't yet been loaded (thus the race condition). And so, no such context for calling into Java is available at that time.

**Comment by jcortell68** [comment]
> Yes unfortunately so, often mandated by the system only having those properties for necessary settings. That's how I know that the properties are unreliable because there are cases where it will break (e.g. people sometimes run into problems where properties aren't unset properly and no vulkan programs work anymore on the device).

Absolutely. It's quite hackish because properties are global but they're being used to control per-process behavior.

**Comment by jcortell68** [comment]
I found a possible solution but it requires API 27 and greater. So not sure if that's a deal breaker.

You can configure a debuggable apk (which the replay server is) to launch with custom environment variables by packaging a wrap shell script in the apk.
https://developer.android.com/ndk/guides/wrap-script

Thoughts?

**Comment by baldurk** [comment]
As long as that won't break on any other android versions and won't cause any other problems that sounds OK, though it seems a bit lightly documented so I'm a little worried it's going to be unpredictable.

**Comment by jcortell68** [comment]
> As long as that won't break on any other android version

Well, that's my point., It requires API 27 and higher, which means if we go that route, RenderDoc will stop working on Android versions older than 8.1. For reference, 8.1 was released at the end of 2017. This requires a judgment call from you, as I don't know how far back you want to support Android, nor do I know the oldest version of Android RenderDoc currently supports.

**Comment by baldurk** [comment]
I didn't want to assume either way, the page you linked isn't clear to me on how the wrap script is actually invoked/enabled/configured, just what its contents should be. If there's no configuration and just any script named `wrap.sh` in a debuggable package is invoked automatically, in that case it sounds like older versions of android would still work fine because they just won't run the script?

If there is a requirement to break compatibility in such a way that older android versions would not work, then no that solution won't be OK. RenderDoc supports Android 6.0 and that must continue to work.

**Comment by jcortell68** [comment]
> older versions of android would still work fine because they just won't run the script?

But the solution would involve two things:
1. make use of the script to set an environment variable
2. change the am-i-in-the-replay server check in libVkLayer_GLES_RenderDoc.so to check the environment variable **instead of** doing a dlsym(marker)

It's (2) that would break older Android versions given the script won't run pre 8.1

**Comment by jcortell68** [comment]
> RenderDoc supports Android 6.0 and that must continue to work.

That's what I needed to know (to avoid spending time on creating a PR). I suppose you can close this as will-not-fix and the bridge will be crossed when it's crossed down the road, if at all.

**Comment by baldurk** [comment]
I may still be missing something then, why would you need to change the check to be instead-of? Why can't both be checked and either one be sufficient? We only want to know that we're being loaded in a replay application, the dlsym may have a "false negative" in the case of Android being broken as in this issue, and the environment variable check may have a false negative if running on an older version of Android which doesn't run the wrapper script, but I don't see that to be a problem and I don't see a way for these to produce false positives either.

I'll close this for now as it seems like there's no good solution and especially since this is an Android bug not a RenderDoc bug, but make a note to try and figure out if there's a way to detect this case and show in the UI as a fatal error. If it starts to show up on real devices as a problem then I can try and figure out a solution.

**Comment by jcortell68** [comment]
Well, my assumption is that if we come up with a fix, we'd want a fix that works in all cases for all supported Androids. The approach of changing the logic to check for both the marker symbol and the environment variable would have a gap. A device with Android 8.0 that experiences the loading issues described here would be unfixed by such a solution.

Regardless, the ticket is now closed but the good discussions will remain here. I'm sure if this situation pops up again in the future, this ticket will help.

**Comment by baldurk** [comment]
Ah I see what you mean. Ideally yes, a solution which works on all android versions would be best, but from the previous discussion on the issue it doesn't seem like any such solution exists :smile: . Particularly if we're not even sure if this bug is possible to reproduce on older Androids. FWIW this isn't without precedent, e.g. until Android 10 GL layers are available RenderDoc uses JDWP/interceptor-lib hooking which does not always work on all devices/apps.

Either way let's leave this for now unless someone comes up with a clear portable fix.

**Comment by jcortell68** [comment]
> but from the previous discussion on the issue it doesn't seem like any such solution exists 😄

It sure has been leaning that way, but there's always a chance there's an epiphany around the corner. 

And yeah, I agree with the open questions of...why exactly is the apk order loading different for our Android platform? Are we alone? If we're not, what range of Android versions are at play? So yeah, probably best to park this until we have a more pressing reason to address it globally. I already have a local fix (which uses the executable filename check you once had; the problematic scenario you said it suffers from is N/A for my company).

And yeah...thank god for Android GLES layers now. Things were not pretty before then.

**Comment by ShabbyX** [comment]
Could we please reopen this issue? Pixel 6 is affected by this, and it seriously hampers usage of RenderDoc.

And about a solution, assuming switching to API 27 does in fact solve the problem, how about at least providing a build config? The official build could continue to use the old API version as long as Android 6 is supported, but at least we could build RenderDoc from source, specify that we are ok with a newer API version, and not have this problem anymore.

**Comment by jcortell68** [comment]
Curious. What version of Android is on your Pixel 6? If this indeed ends up being an issue for only recent Android versions, then Baldur's recommendation to add the new type of check to the existing one, instead of replacing it, would be the cleanest fix. Until now, it wasn't even clear if anyone else but me was seeing this behavior. We now have at least one other person, so yeah...we need to start thinking more seriously about fixing it (as best we can).

**Comment by baldurk** [comment]
I was worried this Android bug wouldn't be isolated to a single case :(.

Yes I think the best solution for this is to allow the use of the wrapper script on newer Androids, use normal detection on older Androids (or I guess newer Androids will use both implicitly). I'll need to read more documentation on the scripts to make sure there aren't any unpleasant side-effects or other bugs. Hopefully this bug is only introduced in a new version of Android so there's no gap in coverage, but at least that will fix it for the common case of broken new Android versions.

**Comment by ShabbyX** [comment]
> What version of Android is on your Pixel 6?

I'm on something more recent now (and still an issue), but this was observable even with the original Android version that shipped with Pixel 6.

Infinite gratitude if it can be fixed!

**Comment by baldurk** [comment]
That commit adds a wrap script to export an env var, and checks it optionally with the proper symbol. I don't have a device with a broken version of Android though so I can't properly test that it fixes the problem (maybe the renderdoc library will still get loaded before the wrap script executes perhaps?).

Can you test to see if that commit fixes it, either with a local build or else with the next nightly?

**Comment by ShabbyX** [comment]
Either I'm getting super lucky suddenly, or indeed the commit fixes it

:1st_place_medal: for @jcortell68 for diagnosing this and :1st_place_medal: for Baldur for the quick fix. Thank you!

**Comment by jcortell68** [comment]
Just got back from PTO. I'll try the fix locally (after disabling the alternative fix) later today and report back

**Comment by jcortell68** [comment]
Baldur, interesting results from trying the fix.

It works, which is good. The curious part is that the fix has altered how the replay server's dynamic linking plays out. I now see libart.so getting loaded early on, which is something that wasn't happening without the wrapper script. The end result is that the race condition no longer happens. Consequently, in logcat, I see the following every time

```
08-18 11:18:05.619  2650  2650 I renderdoc: @62fe663d00000001@ RDOC 002650: [11:18:05] 	android_hook.cpp( 683) - Log 	- Detecting symbol renderdoc__replay__marker by dlsym: yes
08-18 11:18:05.620  2650  2650 I renderdoc: @62fe663d00000002@ RDOC 002650: [11:18:05] 	android_hook.cpp( 684) - Log 	- Detecting symbol renderdoc__replay__marker by getenv: yes
```

What this means is that the setting of the env variable isn't actually needed, which is a bit ironic. The purpose of the change is to introduce an environment variable that can be used to detect the replay server context, but the mechanism used makes the environment variable unnecessary :-)

Bottom line, though...the change fixes the problem, and that's what's important.

**Comment by bigPillowOMG** [comment]
Baldur & jcortell68,  I can also reproduce this issue with OPPO find X3 pro, and luckily your patch works well on my devices,
Thank you~

**Comment by baldurk** [comment]
It's probably going to break again, especially given the previous reply that it is _already_ being unreliable, even just with the workaround to an Android bug. It's impossible to predict how or where though, Android is such a completely terrible and fundamentally broken platform that nothing ever works reliably as it should, and there is always some incredibly stupid new breakage.

My guess is there will be devices where the original android bug manifests but those devices have disabled wrapper scripts so the environment variable workaround doesn't run. It's not feasible to try and predict that though so we'll need to wait and see what goes wrong.

**Comment by bigPillowOMG** [comment]
> It's probably going to break again, especially given the previous reply that it is _already_ being unreliable, even just with the workaround to an Android bug. It's impossible to predict how or where though, Android is such a completely terrible and fundamentally broken platform that nothing ever works reliably as it should, and there is always some incredibly stupid new breakage.
> 
> My guess is there will be devices where the original android bug manifests but those devices have disabled wrapper scripts so the environment variable workaround doesn't run. It's not feasible to try and predict that though so we'll need to wait and see what goes wrong.

Yes, I hava done more tests to verify this patch and found out it does not always work on my devices.
There is  about 50% repetition that I can still reproduce this issue with this patch.

**Comment by bigPillowOMG** [comment]
> Baldur, interesting results from trying the fix.
> 
> It works, which is good. The curious part is that the fix has altered how the replay server's dynamic linking plays out. I now see libart.so getting loaded early on, which is something that wasn't happening without the wrapper script. The end result is that the race condition no longer happens. Consequently, in logcat, I see the following every time
> 
> ```
> 08-18 11:18:05.619  2650  2650 I renderdoc: @62fe663d00000001@ RDOC 002650: [11:18:05] 	android_hook.cpp( 683) - Log 	- Detecting symbol renderdoc__replay__marker by dlsym: yes
> 08-18 11:18:05.620  2650  2650 I renderdoc: @62fe663d00000002@ RDOC 002650: [11:18:05] 	android_hook.cpp( 684) - Log 	- Detecting symbol renderdoc__replay__marker by getenv: yes
> ```
> 
> What this means is that the setting of the env variable isn't actually needed, which is a bit ironic. The purpose of the change is to introduce an environment variable that can be used to detect the replay server context, but the mechanism used makes the environment variable unnecessary :-)
> 
> Bottom line, though...the change fixes the problem, and that's what's important.

I am a little confused that why this patch can change the order of *.so loading....

**Comment by baldurk** [comment]
And it only took _4 hours_ for Android to prove itself to be a garbage platform once again.

Can you share a clean logcat from just starting the remote server - clear the logcat buffer before with the server not running, then connect to that replay context in the UI and let the server start itself and save the logcat.

**Comment by bigPillowOMG** [comment]
> And it only took _4 hours_ for Android to prove itself to be a garbage platform once again.
> 
> Can you share a clean logcat from just starting the remote server - clear the logcat buffer before with the server not running, then connect to that replay context in the UI and let the server start itself and save the logcat.

I have shared the logcat via E-mail(baldurk(at)baldurk.org), please check your Email inbox..

**The tag logs in the logcat is,**
08-22 21:21:18.077 15230 32488 D linker  : dlsym(handle=0x0("n/a"), sym_name="renderdoc__replay__marker", sym_ver="(null)", caller="/data/app/~~MbSMXCHatqRNsIKIhnqgyg==/org.renderdoc.renderdoccmd.arm64-5hiRyGriAn8KPXKX_-e15w==/lib/arm64/libVkLayer_GLES_RDC_Test.so", caller_ns=classloader-namespace@0x7a58b71e80) ...
08-22 21:21:18.077 15230 32488 D linker  : undefined symbol: renderdoc__replay__marker
08-22 21:21:18.077 15230 32488 D linker  : ... dlsym failed: undefined symbol: renderdoc__replay__marker
08-22 21:21:18.077 15230 32488 D linker  : dlerror set to "undefined symbol: renderdoc__replay__marker"
08-22 21:21:18.077 15230 32488 I renderdoc: @630382ce00000001@ RDOC 015230: [21:21:18]     android_hook.cpp( 683) - Log     - Detecting symbol renderdoc__replay__marker by dlsym: no
08-22 21:21:18.077 15230 32488 I renderdoc: @630382ce00000002@ RDOC 015230: [21:21:18]     android_hook.cpp( 684) - Log     - Detecting symbol renderdoc__replay__marker by getenv: no
08-22 21:21:18.078 15230 32488 I renderdoc: @630382ce00000003@ RDOC 015230: [21:21:18]    posix_network.cpp( 429) - Log     - Created and bind socket: 233

**Comment by baldurk** [comment]
What exactly is this build? The log snippet you posted references a file `libVkLayer_GLES_RDC_Test.so` in the `org.renderdoc.renderdoccmd.arm64` package which has never existed and does not exist now, and the log indicates the build is from v1.20 at commit 19b25d86d9247a134cd84731fe4cf9b624633ff5 which is the wrong version and that commit does not exist on this repository.

In fact the log shows two processes starting, one is the RenderDoc remote server which is correctly detected and the second is a captured program.

Please explain what is happening here, and if you can reproduce this issue as you said above please share a logcat *as I requested above* - cleared before the bug appears, and only showing the launch of the remote server when it is misidentified as a captured program without then launching a program for capture. If the bug doesn't manifest, close everything and clear the logcat then start again so you can get a clean log.

**Comment by bigPillowOMG** [comment]
As we all know, RenderDoc is a very excellent graphic debug tool, I hava added some new functions to enhance the compatibility of RenderDoc which is  based on v1.20, such as to support GL_TEXTURE_EXTERNAL_OES textures in OpenGL ES for Android platforms and change the name of libVkLayer_GLES_RenderDoc .so to libVkLayer_GLES_RDC_Test.so.

The commit of 19b25d86d9247a134cd84731fe4cf9b624633ff5 is the "wrap script patch", I just  changing the commit ID for my local git repository.
All in all, those commits are nothing to do with this issue.

Back on this issue itself, I don't know why that I have tested many times but I cannot reproduce this issue again on my device today.... So, the patch works today, feel free to close this case.

**Comment by baldurk** [comment]
In future do not waste my time again by reporting bugs based on a customised non-standard build.
